files/image_config/corefile_uploader/core_uploader.py

```diff
@@ -0,0 +1,294 @@
+
+        os.chdir(INIT_CWD)
+
+    @staticmethod
+    def on_any_event(event):
+        if event.is_directory:
+            return None
+
+        elif event.event_type == 'created':
+            # Take any action here when a file is first created.
+            log_debug("Received create event - " +  event.src_path)
+            Handler.handle_file(event.src_path)
+
+
+    @staticmethod
+    def wait_for_file_write_complete(path):
+        ct_size = -1
+
+        while ct_size != os.path.getsize(path):
+            ct_size = os.path.getsize(path)
+            time.sleep(2)
+
+        time.sleep(2)
+        if ct_size != os.path.getsize(path):
+            raise Exception("Dump file creation is too slow: " + path)
+
+        log_debug("File write complete - " +  path)
+
+
+    @staticmethod
+    def handle_file(path):
+
+        Handler.wait_for_file_write_complete(path)
+
+        lpath = "/".join(cwd)
+        make_new_dir(lpath)
+        os.chdir(lpath)
+
+        # Create a new archive with core & more.
+        metafiles = cfg.get_dict()["metadata_files_in_archive"]
+
+        fname = os.path.basename(path)
+        tarf_name = fname + ".tar.gz"
+
+        cfg.get_core_info(path, hostname)
+        
+        tar = tarfile.open(tarf_name, "w:gz")
+        for e in metafiles:
+            tar.add(metafiles[e])
+        tar.add(path)
+        tar.close()
+        log_debug("Tar file for upload created: " + tarf_name)
+
+        Handler.upload_file(tarf_name, tarf_name)
+
+        log_debug("File uploaded - " +  path)
+        os.chdir(INIT_CWD)
+
+    @staticmethod
+    def upload_file(fname, fpath):
+        daemonname = fname.split(".")[0]
+        i = 0
+        fail_msg = ""
+        
+        while i <= MAX_RETRIES:
+            try:
+                svc = FileService(account_name=acctname, account_key=acctkey)
+
+                l = [sonicversion, asicname, daemonname, hostname]
+                e = []
+                while len(e) != len(l):
+                    e.append(l[len(e)])
+                    svc.create_directory(sharename, "/".join(e))
+
+                log_debug("Remote dir created: " + "/".join(e))
+
+                svc.create_file_from_path(sharename, "/".join(l), fname, fpath)
+                log_debug("Remote file created: name{} path{}".format(fname, fpath))
+                break
+
+            except Exception as e:
+                log_err("core uploader failed: Failed during upload (" + str(e) +")")
+                fail_msg = str(e)
+                i += 1
+                time.sleep(PAUSE_ON_FAIL)
+
+        if i >= MAX_RETRIES:
+            raise Exception("Failed while uploading. msg(" + fail_msg + ") after " + str(i) + " retries")
+
+    @staticmethod
+    def scan():
+        for e in os.listdir(CORE_FILE_PATH):
+            fl = CORE_FILE_PATH + e
+            if os.path.isfile(fl):
+                Handler.handle_file(fl)
```

Comment:
Should core file be removed after a successful uploading? Otherwise every time the service starts, all existing core file will be uploaded again.

Reply:
Reasons:
1) I am not attaching core file to ICM now, so as not to increase ICM DB size,. Hence we need it here.
2) Re-upload is not a problem, as the analyzer would understand that it is duplicate and will not file any ticket
3) If the file is  still in azure storage, it would only get re-updated and no duplicates.
4) In a normal system, the service restart should be pretty seldom (only upon reboot)


On a side note: FYI:
5) We have an upcoming core-mgmt service from Broadcom that would limit the count of core files for a process